### PR TITLE
config-reloader: add the reload-timeout argument

### LINKED ISSUE #5346: Timeout in config-reloader not configurable
**What is missing?**

In `config-reloader` tool, the default timeout of 30 seconds when calling `/-/reload` endpoint of Prometheus is not configurable.

**Why do we need it?**

We currently face an issue where Prometheus takes longer than 30 seconds to reload its configuration.

For exemple :
> caller=main.go:1234 level=info msg="Completed loading of configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml totalDuration=34.152274791s db_storage=2.48µs remote_storage=4.66µs web_handler=660ns query_engine=1.03µs scrape=1.68919009s scrape_sd=89.964915ms notify=27.572µs notify_sd=378.805µs rules=58.754µs tracing=6.2µs

And config-reloader tries and tries again until Prometheus responds in less than 30 seconds thus creating a huge load on it.

<details><summary>See config-reloader logs</summary>

> level=info caller=reloader.go:374 msg="Reload triggered" cfg_in=/etc/prometheus/config/prometheus.yaml.gz cfg_out=/etc/prometheus/config_out/prometheus.env.yaml watched_dirs=/etc/prometheus/rules/prometheus-prm-prometheus-rulefiles-0
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded"
> level=error caller=reloader.go:384 msg="Failed to trigger reload. Retrying." err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=error caller=runutil.go:100 msg="function failed. Retrying in next tick" err="trigger reload: reload request failed: Post \"http://127.0.0.1:9090/-/reload\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
> level=info caller=reloader.go:374 msg="Reload triggered" cfg_in=/etc/prometheus/config/prometheus.yaml.gz cfg_out=/etc/prometheus/config_out/prometheus.env.yaml watched_dirs=/etc/prometheus/rules/prometheus-prm-prometheus-rulefiles-0

</details>

**Environment**

* Prometheus Operator version:

    `v0.59.1`

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by JHeilCoveo**
This new argument enables you to increase (or decrease) the duration for config-reloader to wait after Prometheus' reload.

Fixes #5346

## Description

Allow the time waited for Prometheus' configuration reload to be customized. 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

```release-note
- Add a new parameter `--reload-timeout` to prometheus-config-reloader to customize the wait time of prometheus' configuration reload.
```

### CODE REVIEW

**Review by JoaoBraveCoding — COMMENTED** [review]
LGTM 👍 Just have one question @JHeilCoveo have you validated that you can use the [containers](https://prometheus-operator.dev/docs/operator/strategic-merge-patch/#merging-patch-for-prometheus) field to correctly set the timeout you desired?

**Comment by JHeilCoveo** [comment]
> LGTM +1 Just have one question @JHeilCoveo have you validated that you can use the [containers](https://prometheus-operator.dev/docs/operator/strategic-merge-patch/#merging-patch-for-prometheus) field to correctly set the timeout you desired?

I've used the `command` attribute of `container` to be able to add an argument without remove the ones generated by prometheus-operator.

Like this :
```yaml
spec:
  containers:
  - command:
    - /bin/prometheus-config-reloader
    - --reload-timeout=1m
  # ... snip ...

File: cmd/prometheus-config-reloader/main.go [hunk 1]
```diff
@@ -59,6 +60,7 @@ func main() {
 	watchInterval := app.Flag("watch-interval", "how often the reloader re-reads the configuration file and directories; when set to 0, the program runs only once and exits").Default(defaultWatchInterval.String()).Duration()
 	delayInterval := app.Flag("delay-interval", "how long the reloader waits before reloading after it has detected a change").Default(defaultDelayInterval.String()).Duration()
 	retryInterval := app.Flag("retry-interval", "how long the reloader waits before retrying in case the endpoint returned an error").Default(defaultRetryInterval.String()).Duration()
+	reloadTimeout := app.Flag("reload-timeout", "how long the reloader waits for a response from reload URL").Default(defaultReloadTimeout.String()).Duration()
```
Comment (simonpasquier) [inline cmd/prometheus-config-reloader/main.go:63]:
```suggestion
	reloadTimeout := app.Flag("reload-timeout", "how long the reloader waits for a response from the reload URL").Default(defaultReloadTimeout.String()).Duration()
```

**Review by simonpasquier — COMMENTED** [review]
just one small not otherwise lgtm

**Review by simonpasquier — APPROVED** [review]

**Comment by simonpasquier** [comment]
Thanks!

**Pull request merged**